nRF: radio DFE (and other DT) cleanup

mbolivar-nordic opened the pull request:
The main purpose of this pull request is to audit the nRF SoC .dtsi files and ensure they all have any missing RADIO, FICR, UICR, and EGU nodes.

Adding the missing RADIO nodes in particular is a prerequisite for implementing #30708, which will add a `fem` property to the radio nodes that will allow for configuring the PA/LNA hardware. The rest are just cleanups while I'm here.

While doing this, I noticed that I couldn't really add RADIO nodes for SoCs that don't support the Bluetooth Direction Finding Extension (DFE). This is because all the DFE related properties in the RADIO nodes are marked `required`, which in my opinion doesn't make much sense to add to e.g. the nRF52840 .dtsi file, since that SoC doesn't support DFE.

Therefore, as prep work for the prep work, this PR also reorganizes the DFE support by making all the related properties optional, and adds a new `dfe-supported` property that can be used to test if a given RADIO peripheral supports the extension. That requires some changes to the related controller implementation and bindings documentation, so I took the opportunity to try to do some other docs improvements as well.

mbolivar-nordic committed:
dts: nrf51822: sort /soc nodes

Sort by unit address.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: nrf51822: unify /chosen node settings

There's no need for two separate places for this.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: nrf52805: sort /soc nodes

Sort by unit address.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: nrf52810: sort /soc nodes

Sort by unit address.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: nrf51822: add some missing nodes

Add missing FICR, UICR, RADIO nodes. We already have compatibles for
these defined, and the peripherals are present, so define them for
consistency across SoCs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
soc: nordic_nrf: add NRF_RADIO base address check

Update validate_base_addresses.c for the RADIO peripheral.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: nrf52805: add some missing nodes

Add missing FICR, UICR, RADIO, and EGU nodes. We already have
compatibles for these defined, and the peripherals are present, so
define them for consistency across SoCs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: nrf52810: add some missing nodes

Add missing RADIO and EGU nodes. We already have compatibles for these
defined, and the peripherals are present, so define them for
consistency across SoCs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: nrf52811: sort /soc nodes

Sort by unit address.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: nrf52811: add some missing nodes

Add missing RADIO and EGU nodes. We already have compatibles for these
defined, and the peripherals are present, so define them for
consistency across SoCs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: nrf52820: sort /soc nodes

Sort by unit address.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: nrf52820: add some missing nodes

Add missing EGU and TIMER nodes. We already have compatibles for these
defined, and the peripherals are present, so define them for
consistency across SoCs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: nrf52832: sort /soc nodes

Sort by unit address.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: nrf52832: add some missing nodes

Add missing RADIO and EGU nodes. We already have compatibles for
these defined, and the peripherals are present, so define them for
consistency across SoCs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: nrf52833: sort /soc nodes

Sort by unit address.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: nrf52840: sort /soc nodes

Sort by unit address.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: nrf52840: add some missing nodes

Add missing RADIO and EGU nodes. We already have compatibles for
these defined, and the peripherals are present, so define them for
consistency across SoCs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: nrf52833: add some missing nodes

Add missing EGU nodes. We already have a compatible for
these defined, and the peripherals are present, so define them for
consistency across SoCs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: nrf5340: sort /soc nodes

Sort by unit address. This involves multiple files since there are
includes and secure/non-secure files to deal with.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: samples: bluetooth: DFE fixes

The documentation for Bluetooth Direction Finding Extension (DFE)
samples has various issues:

- references to 'child' images, which do not exist in mainline zephyr

- invalid RST syntax: there are missing ` characters to end arguments
  to :zephyr_file: roles, creating unintelligible output

- incorrect RST usage:

   - using :zephyr_file: instead of :file: when referring to a file
     that the user must create, creating broken links to nonexistent
     files in the zephyr tree

   - using :code: instead of :kconfig: to refer to kconfig options,
     creating output without links to the help for those options

- redundant or duplicated information

- grammar, typos, various bits and pieces

Clean this up. As part of that, move various common bits and pieces of
information to the devicetree bindings index so they can just be
linked to from the sample docs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>
Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

mbolivar-nordic committed:
Bluetooth: Controller: clean up nRF DFE

The way we currently handle direction finding extension (DFE) support
on Nordic nRF5 controllers relies on required devicetree properties
related to DFE in the "nordic,nrf-radio" node.

That doesn't make sense on radios without DFE support, though.
Any .dtsi for an SoC without DFE support which has such a node would
require extraneous DFE related properties like dfe-antenna-num.

Instead of making the properties required, mark them optional. We
indicate the presence of DFE support via a new 'dfe-supported' boolean
property which the SoC .dtsi files can set (or not) depending on
support.

This gives us the opportunity to do some cleanup in the Kconfig,
removing CONFIG_HAS_HW_NRF_RADIO_BLE_DF since we know from the
devicetree whether DFE support is available.

Handle that change appropriately in radio_df.c. This gives us an
opportunity to improve readability in the devicetree-related macro
magic in that file.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
soc: nordic_nrf: add NRF_DT_GPIOS_TO_PSEL_OR

Helper macro for getting the PSEL value from optional devicetree
properties.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

ppryga commented on dts/arm/nordic/nrf52811.dtsi:
```diff
@@ -50,6 +50,14 @@
+			reg = <0x40001000 0x1000>;
+			interrupts = <1 NRF_DEFAULT_IRQ_PRIORITY>;
+			status = "okay";
+			dfe-supported;
```
If DFE is added here, add also information about nRF52811DK in DF samples `README.rst`, please:
- https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/bluetooth/direction_finding_connectionless_rx/README.rst#requirements
- https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/bluetooth/direction_finding_connectionless_tx/README.rst#requirements

mbolivar-nordic replied:
> If DFE is added here, add also information about nRF52811DK in DF samples `README.rst`, please:

Can we instead say "nRF SoC with DFE support" or something like that? This way it is is more future-proof. We can tell users that they can check their devicetree for 'dfe-supported' in the radio node or check the product specification if they are unsur.

ppryga replied:
>"nRF SoC with DFE support"
Good idea!

mbolivar-nordic commented on dts/bindings/net/wireless/nordic,nrf-radio.yaml:
```diff
@@ -2,21 +2,54 @@
+    This also means that, for example, when using four GPIOs, the pattern count
+    cannot be greater than 16, and the maximum allowed value is 15.
+
+    If the number of switch-sample periods is greater than the number of stored
+    switching patterns, the radio loops back to the first pattern.
```
> Which is better in your opinion?

More context would be good, I think.

> BTW. The context was missing in former version of README.rst.

Thanks for making me feel better about not understanding it well when I re-read it :).

ppryga replied:
Ok, I'm fine with extending context here. Do you need any support in that?

mbolivar-nordic replied:
> Do you need any support in that?

If you have suggested changes I would be happy to apply them. Otherwise I will take a stab at it.

mbolivar-nordic commented:
v2: I moved the definition of `PDU_ANTENNA` (which is the `dfe-pdu-antenna` property value if set, or 0 if undefined) outside of this conditional compilation check:

```
#if defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX) || \
	defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_RX)
```

I also changed the BUILD_ASSERT condition to `BUILD_ASSERT(HAS_PDU_ANTENNA, ...)`, making the property required to use this file.

@ppryga I did this because of CI build errors related to `PDU_ANTENNA` being undefined in `radio_df_pdu_antenna_switch_pattern_get()` and `radio_df_ant_switch_pattern_set()`. Since these are not conditionally defined based on the above Kconfigs, it seems the property really is always required if radio_df.c is going to work.

I think we just missed this because `dfe-pdu-antenna` was previously `required`.

mbolivar-nordic commented:
v3 silences a spurious compliance warning by playing with whitespace.

mbolivar-nordic commented:
v4:

- attempt to address documentation requests from @ppryga (please revisit)
- make a final simplification to the macro usage in radio_df.c

Note that I can only build test the DFE bits.

ppryga commented:
> v2: I moved the definition of `PDU_ANTENNA` (which is the `dfe-pdu-antenna` property value if set, or 0 if undefined) outside of this conditional compilation check:
> 
> ```
> #if defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX) || \
> 	defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_RX)
> ```
> 
> I also changed the BUILD_ASSERT condition to `BUILD_ASSERT(HAS_PDU_ANTENNA, ...)`, making the property required to use this file.
> 
> @ppryga I did this because of CI build errors related to `PDU_ANTENNA` being undefined in `radio_df_pdu_antenna_switch_pattern_get()` and `radio_df_ant_switch_pattern_set()`. Since these are not conditionally defined based on the above Kconfigs, it seems the property really is always required if radio_df.c is going to work.
> 
> I think we just missed this because `dfe-pdu-antenna` was previously `required`.

`radio_df_pdu_antenna_switch_pattern_get()` and `radio_df_ant_switch_pattern_set()` are referenced only if `CONFIG_BT_CTLR_DF_ANT_SWITCH_TX` or `CONFIG_BT_CTLR_DF_ANT_SWITCH_RX` is defined. I didn't put them in conditional compilation because in condition mentioned before they are be removed during LTO.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -14,90 +14,103 @@
+ * Check that we have an antenna switch pattern.
+ */
+
+#define HAS_PDU_ANTENNA DT_NODE_HAS_PROP(RADIO_NODE, dfe_pdu_antenna)
+
+#define PDU_ANTENNA DT_PROP_OR(RADIO_NODE, dfe_pdu_antenna, 0)
+
+BUILD_ASSERT(HAS_PDU_ANTENNA,
+	     "Missing antenna pattern used to select antenna for PDU Tx. "
+	     "Set the dfe-pdu-antenna devicetree property.");
```
Move it back into 
```
#if defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX) || \
	defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_RX)
```
please. 

Also make `radio_df_pdu_antenna_switch_pattern_get()` and `radio_df_ant_switch_pattern_set()` functions conditionally compilable. It will decrease default compilation code coverage but it will give possibility to build DF beacon application working in AoA mode. The APP will not need DTS overlay for DFE configuration.
If `BUILD_ASSERT` is not conditional code, even simple app, transmitting CTE in AoA mode (that does not use antenna switching feature of DFE), requires an DTS overlay.

mbolivar-nordic replied:
> Also make `radio_df_pdu_antenna_switch_pattern_get()` and `radio_df_ant_switch_pattern_set()` functions conditionally compilable.

OK, I will put them under the same ifdef as the BUILD_ASSERT.

However, usage of these functions is different and I'm not sure that is always correct.

- radio_df_pdu_antenna_switch_pattern_set() is called from lll_df.c, and only if CONFIG_BT_CTLR_DF_ANT_SWITCH_TX=y or CONFIG_BT_CTLR_DF_ANT_SWITCH_RX=y
- radio_df_pdu_antenna_switch_pattern_get() is only called from lll_sync.c, but only if CONFIG_BT_CTLR_DF_SCAN_CTE_RX=y

As far as I can tell, CONFIG_BT_CTLR_DF_SCAN_CTE_RX doesn't depend on either CONFIG_BT_CTLR_DF_ANT_SWITCH_TX or CONFIG_BT_CTLR_DF_ANT_SWITCH_RX. So I'm not sure lll_sync.c is doing the right thing there.

Is this actually OK?

mbolivar-nordic replied:
>>Also make radio_df_pdu_antenna_switch_pattern_get() and radio_df_ant_switch_pattern_set() functions conditionally compilable.

> I'm not sure lll_sync.c is doing the right thing there.

Yep, getting a link error in CI:

```
/workdir/zephyr/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:400: undefined reference to `radio_df_pdu_antenna_switch_pattern_get'
```

I'm not sure what changes you'd like to see here in terms of making radio_df_pdu_antenna_switch_pattern_get() conditionally compiled.

ppryga replied:
Ah... I'm sorry. In such situation move the the radio_df_pdu_antenna_switch_pattern_get function out of `#if defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX) || defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_RX)` conditional compilation block and change like this:
```
uint8_t radio_df_pdu_antenna_switch_pattern_get(void)
{
#if defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX) || \
	defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_RX)
	return PDU_ANTENNA;
#else
	return 0;
#endif
}
```

mbolivar-nordic replied:
OK, understood. I'd like to suggest a slightly different approach: moving both of these lines outside of the conditional compilation block:

```
#define PDU_ANTENNA DT_PROP_OR(RADIO_NODE, dfe_pdu_antenna, 0)

uint8_t radio_df_pdu_antenna_switch_pattern_get(void)
{
	return PDU_ANTENNA;
}
```

I think that's a little easier to read. LMK if you disagree.

ppryga replied:
Yes, it looks better. I'll remember to use `DT_PROP_OR` in future for such cases.

ppryga-nordic reviewed (changes requested):
I've added some suggestions related with DFE documentation in `nrf-radio.yaml`. Besides that two small things in radio_df.c.
I'll test changes with DF samples when you update the PR.

mbolivar-nordic commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -14,90 +14,103 @@
-#define DFE_GPIO_ALLOWED_ANT_NUM COND_CODE_1(DFE_GPIO_NUM_IS_ZERO,       \
-					     (BIT(DFE_GPIO_NUM)), (0))
+/*
+ * Check that we have an antenna switch pattern.
```
>     * guard, let's say it is kind of a ramp up for Constant Tone reception or transmission (`SWITCHPATTERN[0]`),

Is [0] a typo? Below there is a comment that says SWITCHPATTERN[1] is for the guard and reference period.

ppryga replied:
I'm sorry, there should be SWITCHPATTERN[1]. I've verified that in Radio documentation.

mbolivar-nordic commented:
v5:

- rebased to address dtsi conflicts
- added more DT bindings docs
- reworked dfe-dpu-antenna property handling
- other comments and documentation fixes

ppryga commented on dts/bindings/net/wireless/nordic,nrf-radio.yaml:
```diff
@@ -2,21 +2,83 @@
+      In AoA mode, DFE collects reference IQ samples. For this state,
+      DFE selects the antenna by applying SWITCHPATTERN[1].
+
+    * Switch-sample, in this state actual antenna switching happens.
```
nit:
```suggestion
    * Switch-sample: in this state actual antenna switching happens.
```

ppryga-nordic reviewed (changes requested):
I'm sorry that you have to apply comments about my other comments...

mbolivar-nordic commented:
v6:

- bindings documentation fixes
- fix link error by making radio_df_pdu_antenna_switch_pattern_get() always defined

mbolivar-nordic commented:
@ppryga thanks for the review and for re-checking everything. I've applied your suggestions and resolved comments. I left one comment open where I took a slightly different approach from your suggestion.

ppryga commented:
@mbolivar-nordic I'll check samples if they are working and then approve the PR.

mbolivar-nordic commented:
> @mbolivar-nordic I'll check samples if they are working and then approve the PR.

@ppryga ping. I've rebased on top of your latest changes and CI is green.

ppryga-nordic reviewed (approved):
Samples are working. Thanks for taking care about it.

hubertmis commented on dts/bindings/net/wireless/nordic,nrf-radio.yaml:
```diff
@@ -29,94 +92,91 @@ properties:
     interrupts:
       required: true
 
+    dfe-supported:
```
This binding should also include `802154-supported` or something like this, because only some radios support this protocol, namely nRF52811, nRF52820, nRF52833, nRF52840, nRF5340.

hubertmis commented on soc/arm/nordic_nrf/Kconfig.peripherals:
```diff
@@ -133,9 +133,6 @@ config HAS_HW_NRF_RADIO_BLE_CODED
 config HAS_HW_NRF_RADIO_IEEE802154
```
This Kconfig option should be replaced by an entry from DT indicating if given SoC supports IEEE 802.15.4 protocol

hubertmis reviewed (commented):
nRF radio DT binding being introduced in this PR should include property indicating if given SoC supports IEEE 802.15.4.

I'm aware it's not necessary for the goal of this PR. We may introduce this property later.

mbolivar-nordic commented:
> We may introduce this property later.

I will prepare another PR later, then. This one has been open for some time now and I'd like to continue work on the FEM bindings it unblocks.

> nRF52811, nRF52820, nRF52833, nRF52840, nRF5340.

Thanks for compiling the list of SoCs :+1:

cvinayak commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -31,7 +33,8 @@ config BT_LLL_VENDOR_NORDIC
 	select BT_CTLR_ADV_ISO_SUPPORT
 	select BT_CTLR_SYNC_ISO_SUPPORT
-	select BT_CTLR_DF_SUPPORT if HAS_HW_NRF_RADIO_BLE_DF
+	select BT_CTLR_DF_SUPPORT if \
+		$(dt_node_has_bool_prop,$(DT_PATH_NORDIC_RADIO),dfe-supported)
```
```suggestion
	select BT_CTLR_DF_SUPPORT if DT_NORDIC_RADIO_DF_SUPPORT
```

cvinayak commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -5,6 +5,8 @@
 
 if BT_LL_SW_SPLIT
 
+DT_PATH_NORDIC_RADIO := $(dt_nodelabel_path,radio)
+
```
```suggestion
DT_PATH_NORDIC_RADIO := $(dt_nodelabel_path,radio)
DT_NORDIC_RADIO_DF_SUPPORT := $(dt_node_has_bool_prop,$(DT_PATH_NORDIC_RADIO),dfe-supported)
```

cvinayak reviewed (changes requested):
Did only a quick review pass through of dt files, no comments. A minor change requested which I am not sure if semantically correct, but will appear cleaner.

Otherwise, rest looks good to me.

mbolivar-nordic commented:
Thanks @cvinayak. Something like that is indeed possible; please revisit.
